### PR TITLE
HRINT-2442 [Demo]  Highlighting - Map Index

### PR DESCRIPTION
**Pull request by Danielle9897**

### CODE REVIEW

**Comment by Danielle9897** [comment]
@DejanMilicic can you pls review ?

File: DemoServer/Controllers/Demos/TextSearch/HighlightResultsMapIndex/metadata.json [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+            "DescriptionLines": [
+                "* Prior to making the search query, you can define the tags that will mark the searched terms within the text fragments. This is an optional step.",
+                " ",
+                "* _To define your own tags_ create a `HighlightingOptions` object with _PreTags_ & _PostTags_.<br>",
+                "  This object is passed to the query - see next step.",
+                "",
+                "* _The default pre & post tags_ are: `<b style='background:yellow'>......</b>`<br>",
+                "  The color will change per field that is searched.",
+                "",
+                "* When defining multiple tag strings in the tags array,<br> and when querying with multiple highlight field requests,<br> then each tag string is used per field requested respectively."
```
Comment (DejanMilicic) [inline DemoServer/Controllers/Demos/TextSearch/HighlightResultsMapIndex/metadata.json:193]: I cannot comprehend the meaning of this. Could you clarify it somehow, please?

**Review by DejanMilicic — CHANGES_REQUESTED** [review]

File: DemoServer/Controllers/Demos/TextSearch/HighlightResultsMapIndex/HighlightResultsMapIndexController.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                #region Step_4
+                Store(x => x.Title, FieldStorage.Yes);
+                Store(x => x.Notes, FieldStorage.Yes); 
+                
+                Index(x => x.Title, FieldIndexing.Search);
+                Index(x => x.Notes, FieldIndexing.Search);
+                
+                TermVector(x => x.Title, FieldTermVector.WithPositionsAndOffsets);
+                TermVector(x => x.Notes, FieldTermVector.WithPositionsAndOffsets);
+                #endregion
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/TextSearch/HighlightResultsMapIndex/HighlightResultsMapIndexController.cs]: done

File: DemoServer/Controllers/Demos/TextSearch/HighlightResultsMapIndex/HighlightResultsMapIndexController.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            #endregion
+           
+            public EmployeesDetails()
+            {
+                #region Step_3
+                Map = employees => from employee in employees
+                    select new IndexEntry
+                    {
+                         Title = employee.Title,
+                         Notes = employee.Notes[0]
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/TextSearch/HighlightResultsMapIndex/HighlightResultsMapIndexController.cs]: done

File: DemoServer/Controllers/Demos/TextSearch/HighlightResultsMapIndex/metadata.json (hunk 1 above)
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/TextSearch/HighlightResultsMapIndex/metadata.json]: as discussed: removed this bullet + separated demos

File: DemoServer/Controllers/Demos/TextSearch/HighlightResultsMapIndex/HighlightResultsMapIndexController.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            Highlightings titleHighlightings;
+            Highlightings notesHighlightings;
+            
+            #region Demo
+            List<Employee> employeesResults;
+         
+            using (IDocumentSession session = DocumentStoreHolder.Store.OpenSession())
+            {
+                #region Step_5
+                HighlightingOptions tagsToUseInFragment = new HighlightingOptions()
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/TextSearch/HighlightResultsMapIndex/HighlightResultsMapIndexController.cs]: done

**Comment by Danielle9897** [comment]
@DejanMilicic can you pls review ?

File: DemoServer/Controllers/Demos/TextSearch/HighlightQueryResultsMapReduce/HighlightQueryResultsMapReduceController.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                    group result by result.Artist into g
+                    select new IndexEntry
+                    {
+                        Artist = g.Key,
+                        AllSongTitles = string.Join(" ", g.Select(x => x.AllSongTitles))
+                    };
+                #endregion 
+                
+                #region Step_5
+                Store(x => x.Artist, FieldStorage.Yes);
```
Comment (DejanMilicic) [inline DemoServer/Controllers/Demos/TextSearch/HighlightQueryResultsMapReduce/HighlightQueryResultsMapReduceController.cs:60]: is there a need to store artist on the index?

**Review by DejanMilicic — CHANGES_REQUESTED** [review]

File: DemoServer/Controllers/Demos/TextSearch/HighlightQueryResultsMapReduce/HighlightQueryResultsMapReduceController.cs (hunk 5 above)
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/TextSearch/HighlightQueryResultsMapReduce/HighlightQueryResultsMapReduceController.cs:60]: yes - an exception is thrown if you don't

File: DemoServer/Controllers/Demos/TextSearch/HighlightQueryResultsBasics/metadata.json [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+            "DescriptionLines": [
+                "* In order to search and get fragments with highlighted terms, the index field on which you search must be configured with:",
+                "  * `Store()`<br>",
+                "     Set _FieldStorage.Yes_ to store the field in the index.",
+                "",
+                "  * `Index()`<br>",
+                "     Set _FieldIndexing.Search_ to allow Full Text Search.",
+                "",
+                "  * `TermVector()`<br>",
+                "     Set _FieldTermVector.WithPositionsAndOffsets_ to store the term's position and offset from the text.",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/TextSearch/HighlightQueryResultsBasics/metadata.json:148]: 1. well, I should change offset to `offsets`
2. from what I see now in https://livebook.manning.com/concept/lucene/offset
`offsets` are 'start and end character position' of the term
3. maybe write: to store the term's position and offsets in the text ?

File: DemoServer/Controllers/Demos/TextSearch/HighlightQueryResultsBasics/metadata.json [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+                    "Type" : "link",
+                    "Url"  : "http://makble.com/what-is-term-vector-in-lucene"
+                }
+            ]
+        },
+        {
+            "Title": "Make a search query",
+            "Slug" : "step-5",
+            "DescriptionLines": [
+                "* Query for documents from collection 'Employees' that contain specific terms.",
```
[the remark this answers is not in the document]
Comment (Danielle9897) [reply DemoServer/Controllers/Demos/TextSearch/HighlightQueryResultsBasics/metadata.json:192]: right - refactored to:
`Query the 'Employees' collection using the defined index for documents that contain the specified term.`

**Review by DejanMilicic — APPROVED** [review]

File: DemoServer/Controllers/Demos/TextSearch/HighlightQueryResultsBasics/metadata.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (DejanMilicic) [reply DemoServer/Controllers/Demos/TextSearch/HighlightQueryResultsBasics/metadata.json]: sure

**Pull request merged**